File: ben/prolog/P31-P41-arithmetic/p41.pl
% A list of Goldbach compositions.

:- ensure_loaded(p40).

goldbach_list(L, U) :- 
    L < 4, 
    goldbach_list(4, U).
goldbach_list(L, U) :- 
    L > U, !.
goldbach_list(L, U) :-
    \+ 0 is L mod 2,
    L1 is L + 1,
    goldbach_list(L1, U).
goldbach_list(L, U) :- 
    0 is L mod 2,
    goldbach(L, [G1, G2]),
    print_goldbach(L, G1, G2),
    L1 is L + 2, 
    goldbach_list(L1, U).

% with limit 
goldbach_list(L, U, M) :- 
    L < 4, 
    goldbach_list(4, U, M).
goldbach_list(L, U, _) :- 
    L > U, !.
goldbach_list(L, U, M) :- 
    \+ 0 is L mod 2, 
    L1 is L + 1, 
    goldbach_list(L1, U, M).
goldbach_list(L, U, M) :- 
    0 is L mod 2, 
    goldbach(L, [G1, G2]), 
    (  G1 > M 
    -> print_goldbach(L, G1, G2)
    ;  true),
    L1 is L + 2, 
    goldbach_list(L1, U, M).

print_goldbach(T, G1, G2) :- 
    writef('%t = %t + %t', [T, G1, G2]), nl.

File: ben/haskell/Q31-41-arithmetic/p41.hs
{-
Given a range of integers by its lower and upper limit, print a list of all even numbers and their Goldbach composition.

In most cases, if an even number is written as the sum of two prime numbers, one of them is very small. Very rarely, the primes are both bigger than say 50. Try to find out how many such cases there are in the range 2..3000.
-}

goldbachList :: Int -> Int -> [(Int, Int)]
goldbachList a b = map goldbach [lower,lower+2..upper]
    where lower = if mod a 2 == 0
                      then a
                  else a + 1
          upper = if mod b 2 == 0
                      then b
                  else b + 1

goldbachList' :: Int -> Int -> Int -> [(Int, Int)]
goldbachList' a b c = filter (\ (x,y) -> x > c) (goldbachList a b)

-- code from p40

goldbach :: Int -> (Int, Int)
goldbach nb = head [(a,b) | a <- eratosthenes nb, 
                            b <- eratosthenes nb,
                            a + b == nb]

-- code from p39

eratosthenes :: Int -> [Int]
eratosthenes nb
    | nb < 2 = []
    | otherwise = removeMultiples [2..nb]

removeMultiples :: [Int] -> [Int]
removeMultiples [] = []
removeMultiples (x:xs) = x : (removeMultiples . filter (\a -> mod a x /= 0) $ xs)
